aten/src/ATen/native/Linear.cpp

```diff
@@ -136,241 +136,333 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+  int perm_index = 0;
+
+  // Start index of ellipsis dimensions in the permuted shape
+  int64_t ell_index = 0;
+  found_ell = false;
+
+  if (arrow_pos == std::string::npos) {
+    // Implicit output is ellipsis (...) + labels seen only once
+    perm_index = ell_num_dim;
+    found_ell = true;
+    for (int label = 0; label < total_labels; ++label) {
+      if (label_count[label] == 1) {
+        label_perm_index[label] = perm_index++;
       }
     }
-    // then the indices that occur exactly once in alphabetic order
-    for (size_t idx = 0; idx < number_of_letters; idx++) {
-      if (num_letter_occurrences[idx] == 1) {
-        idxes_to_preprocessed_dims[letter_mapping[idx]] = num_output_dims;
-        num_output_dims++;
+  } else {
+    // Parse explicit output
+    std::string rhs = equation.substr(arrow_pos + 2);
+    for (std::size_t i = 0; i < rhs.length(); ++i) {
+      switch (rhs[i]) {
+        case ' ':
+          // Ignore spaces
+          break;
+
+        case '.':
+          TORCH_CHECK(
+              // There can only be one ellipsis in the output
+              !found_ell,
+              "einsum() found \'.\' for output but an ellipsis (...) was already found");
+          TORCH_CHECK(
+              // Ensure ellipsis is correct
+              i + 2 < rhs.length() && rhs[++i] == '.' && rhs[++i] == '.',
+              "einsum() found \'.\' for output that is not part of any ellipsis (...)");
+          ell_index = perm_index;
+          perm_index += ell_num_dim;
+          found_ell = true;
+          break;
+
+        default:
+          TORCH_CHECK(
+              rhs[i] >= 'a' && rhs[i] <= 'z',
+              "einsum() subscripts must be in range [a, z] but found ",
+              rhs[i],
+              " for the output");
+          TORCH_CHECK(
+              // Ensure label appeared at least once for some input operand and at
+              // most once for the output
+              label_count[rhs[i] - 'a'] > 0,
+              "einsum() output subscript ",
+              rhs[i],
+              label_count[rhs[i] - 'a'] == -1
+                  ? " appears more than once in the output"
+                  : " does not appear in the equation for any input operand");
+          label_perm_index[rhs[i] - 'a'] = perm_index++;
+          
+          // Set to -1 to mark that this label already appeared in the output
+          label_count[rhs[i] - 'a'] = -1;
       }
     }
   }
-  // now we assign the idxes_to_preprocessed_dims (index -> dimension in preprocessed / output tensors)
-  // for the non-output indices - those that are eventually summed over
-  int64_t position = num_output_dims;
-  for (int64_t i = 0; i < num_total_idxes; i++) {
-    if (idxes_to_preprocessed_dims[i]==-1) {
-      idxes_to_preprocessed_dims[i] = position;
-      position++;
+
+  // Save output size before adding sum dims
+  int out_size = perm_index;
+
+  // If ellipsis is not part of the output, add to contraction dimensions
```

Comment:
If we did not have ellipsis for output, I add it to contraction list to be summed out